skimage/feature/_canny.py

```diff
@@ -275,11 +277,27 @@ def fsmooth(x):
         # Else if high_threshold and/or low_threshold are callables
         # apply them to determine the threshold value / image
         if callable(high_threshold):
-            high_threshold = high_threshold(magnitude)
+            try:
+                high_threshold = high_threshold(magnitude)
+            except:
+                traceback.print_exc()
+                raise ValueError("Callable `high_threshold` raised above exception.\nIt must take one input (image array) and return a scalar value or image array of the same shape as input image")
+
+            if not( isinstance(high_threshold, numbers.Number) or (
+                    isinstance(high_threshold, np.ndarray) and
+                    (high_threshold.shape == magnitude.shape))):
+                raise ValueError("Callable `high_threshold` must take one input (image array) and return a scalar value or image array of the same shape as input image")
```

Comment:
Thanks @jmetz! These are great, except they should be `TypeError`s. (the same error as when a function is called with the wrong arguments.) And incidentally the `except` above should not be bare — it should only catch `TypeError`s!

Reply:
Ah... So I checked about TypeError Vs ValueError, and because the issue is with the "value" of the callables, I thought ValueError would be more appropriate?
I.e. to quote the docs for ValueError:
"Raised when a built-in operation or function receives an argument that has the right type but an inappropriate value, and the situation is not described by a more precise exception such as IndexError."

Also as for not having a bare except (I know that's generally quite bad), if there is any issue with a called function, I thought it might still be useful to have the info from canny also, not just have it raise the internal exception... Happy to only catch ValueError though if that's preferred.

Reply:
(put another way, as we're in the case where eg high_threshold *is* a callable, it's type is technically already correct... Just that it's specific value is wrong... Right?)

Reply:
I once thought the same as you, but I've learned a bit more about type theory since then. =) Specifically, the type of the callable is not just "callable", but `Union[(array -> float), (array -> array)]`. That is: the inputs and outputs of the callable are part of the type. (Whereas a ValueError would stem from, e.g. returning a float that is outside some allowed range, for example.)

And, as I mentioned, you should catch *TypeError*, because that is the error we are detecting: when the function is unhappy with the inputs as provided.

The errors will propagate through if they are of a different type, and this is what we want: our error message that "you have provided the wrong type of function" will not necessarily be accurate if we catch all errors. If someone, for example, provides a threshold function that sends the data to some Google Compute Engine to process on their TPUs (to get back a threshold 😂), and there is a NetworkError, we don't want to say that it's the wrong kind of function: it is the right kind, but it failed for extraneous reasons.

Reply:
@jni - many thanks for the explanation, glad I know more about type theory wrt functions now too :) - I'll amend  in the last commit in the morning.